# RNAseq_Data_process_pipeline
<br>
What customer need to do:<br>
1. The user need to set an empty working directory using mkdir and cd command. (This programme will automatically copy all the files from “/localdisk/data/BPSM/ICA1” to this directory and start analysis.)
2. Download and uncompress the “B222908-2022.ICA1.tar.gz.cpt” file. Encryption key is: 123456. Extract the “tar” file, you will see three files: all_the_things_I_did, FoldChange.sh, script.
3. Type in   ./script    to run the programme. The code will only run for 3 minutes.
4. This programme has calculated the fold change of uninduced and induced groups that are in the same type and time. If you want to do other kinds of "group-wise" comparisons, you can simply type in the following command to run the “FoldChange.sh” programme. You should input file names of two groups that are end with <br>“_expression_average.txt”.
<br>./FoldChange.sh <original_file> <subsequent_file>
<br>For example:
<br>./FoldChange.sh WT_0_Uninduced_expression_average.txt  WT_48_Uninduced_expression_average.txt
<br>
Flowchart that describes how the pipeline processes the data<br>
<img width="56" alt="image" src="https://github.com/Christina002128/RNAseq_Analysis/assets/115002249/db7029ca-a5fc-4c98-98e3-87208d946e3f">

<img width="451" alt="image" src="https://github.com/Christina002128/RNAseq_Analysis/assets/115002249/9d74c851-a114-486a-9be2-3e55d614ead6">
<br>
Results<br>
Results are all in the “all_result_folder”. <br>
1. quality_result/fastfile_result.txt
“quality_result” folder contains files generated from fastqc and a summary file called “ fastfile_result.txt”. 
Part of the table in “ fastfile_result.txt” is shown here:
 <img width="451" alt="image" src="https://github.com/Christina002128/RNAseq_Analysis/assets/115002249/6af33899-c3b7-448c-b8ef-45bfcbf01bfb">

There is no more than one “fail” in each sequence quality test, and the failed item is only “the Per base sequence content”. It means this module will fail if the difference between A and T, or G and C is greater than 20% in any position. Because of the biased selection of random primers, nearly all RNA-Seq libraries will fail this module. But this is not a problem which can be fixed by processing, and it doesn't seem to adversely affect the ability to measure expression. As a result, we can draw a conclusion that all the sequence files are with good quality.
2. *aligned.bed
48 “*_aligned.bed” files contain tables of the expression level of each gene from each paired-end sequence file. The columns represent: 
chromosome , chr_start, chr_end, gene_name, gene_description,  number_of_aligned_reads  separately.
 <img width="452" alt="image" src="https://github.com/Christina002128/RNAseq_Analysis/assets/115002249/c335b924-5e6d-48dd-8468-d4df043fe9b6">

3. *expression.txt
Each “*expression.txt” file contains gene expression level of samples that are in the same group. Columns represent gene_name, gene_description and expression column of each sample.
For example, the table of “WT_0_Uninduced_expression.txt” is shown here:
 <img width="452" alt="image" src="https://github.com/Christina002128/RNAseq_Analysis/assets/115002249/1840486d-e35b-4aa7-9b76-a140894d4077">

4. *expression_average.txt
“*expression_average.txt” files contain the average gene expression level of each group.
For example, the table of “WT_0_Uninduced_expression_average.txt” is shown here:
 <img width="452" alt="image" src="https://github.com/Christina002128/RNAseq_Analysis/assets/115002249/b8e97f00-af06-44fd-828c-8f92feba27e9">

5. *FoldChange.txt
This programme has calculated the fold change of uninduced and induced groups that are in the same type and time. If you want to do other kinds of "group-wise" comparisons, you can simply type in the following command to run the FoldChange.sh programme. You should input file names of two groups that can be inputted are end with “_expression_average.txt”.
./FoldChange.sh <original_file> <subsequent_file>
 For example:
./FoldChange.sh WT_0_Uninduced_expression_average.txt  WT_48_Uninduced_expression_average.txt

<br>
Process Description (highlighted difficulties)
<br><br>1. 
To do quality analysis, I used parallel to accelerate the fastqc process.
<br>“parallel -j 48” is to run 48 jobs at the same time.
<br><br>2. 
To summarize the quality of the sequence, I used the content from summary.txt files generated by fastqc. I calculated the number of passes, fails and warnings and displayed failed items of each sequence file using “for” loop, “if”, “paste” and “grep” commands, and put them into “fastfile_result.txt” file. "echo finished” for the process.
<br><br>3.
To do alignment, I uncompressed the sequence files and genome file using gzip -d command in parallel. Then I chose hisat2 programme to align the read pairs to the Trypanosoma congolense genome because hisat2 is more recent and more efficient than bowtie2, and it also has more features such as spliced read alignment. (But actually, it’s not much difference since we don’t care about intron here.)
<br>First, we need to create index of the genome file, using hisat2-build with 16 threads (-p). Then do hisat to each two pair-end sequence to align them with the genome and generate files in sam format files using “for” loop (hisat2 -p <number of threads> -x <genome index>, -s <output sam_file>, use “&” to parallelize them, and “wait” after the loop). Then convert them to bam format using samtools and sort the content in order using “for” loop and “&” and “wait”. The results are sent to “*_sort.bam” files. "echo finished” for each process.
<br><br>4. 
Convert “*_sorted.bam” files to bed format using bedtools (bedtools bamtobed -i). Count the number of reads that align to the regions of the genome that code for genes (assume all genes have no introns) using “bedtools intersect”. For each feature in the “a” file, the number of overlapping features in the “b” file. Thus, the -a file should be genome, the -b file should be each sequence.bed file. -c is to report the number of hits in “b” file. Also used “for” loop. “&”and “wait” command for parallel.<br>
The results are sent to “*aligned.bed” files. "echo finished” at the end.<br>
<br>5. 
To calculate the statistical mean of expression levels of each gene for each group, I first generated files that contain sample names that are in the same group (in same type, same time, and same treatment). And because “*_aligned.bed ” files are already in order and also in same length, I can simply sum up each line of files that are in the same group, and make an expression table for each group. These are done by three nested “for” loops, and also “while, grep, cat, paste, cut  awk” commands.<br>
Because each group have different number of samples, I used ${count} to record the number of samples that group and used awk to calculate average. <br>
Highlight difficulties: If we want to use an outside variable in awk, we need single quotes: ‘$count’. So, if ‘$count’ equals 3, then sum up 3 columns of expression levels and divided by 3. If ‘$count’ equals 4, then sum up 4 columns of expression levels and divided by 4. If it is not in these two conditions, then print out "error: one file have more than 4 expressions!!”. 
<br>Because these loops generated three files in 0h Induced groups and these groups does not exist, so they are removed at the end of the loops. The results are sent to “*expression_average.txt” files. 
<br><br>6.
To generate "fold change" data for the "group-wise" comparisons, I wrote a bash script called FoldChange.sh to calculate fold change for any two groups using formular: foldechange = (Y-X)/X.  It can generate fold change by inputting an original file(X) and a subsequent file(Y) from all *_expression_average.txt files.
<br>In the BPSM_ICA1_script, it automatically called the FoldChange.sh and generated the fold change of uninduced and induced groups that are in the same type and time.
<br>If the users want to do other kinds of "group-wise" comparisons, they can simply type in:
<br>./FoldChange.sh <original_file> <subsequent_file><br>
File names that can be inputted are end with “_expression_average.txt”. 
<br><br>7.
Move the useful result files into “all_result_folder” directory and copy programme files to this directory. Then remove all unwanted files and folders that generated from the programme.
<br>
